packages/tre/src/storage2/blob/store.ts

```diff
@@ -0,0 +1,191 @@
+  // won't be able to guess the ID, and we don't allow listing blobs.
+  //
+  // For example, if we put a blob in the store, then fail to insert the blob ID
+  // into the SQLite database for some reason during a transaction (e.g.
+  // `onlyIf` condition failed), no other operations can read that blob because
+  // the ID is lost (we'll just background-delete the blob in this case).
+
+  get(
+    id: BlobId,
+    range?: InclusiveRange
+  ): Promise<ReadableStream<Uint8Array> | null>;
+  get(
+    id: BlobId,
+    ranges: InclusiveRange[],
+    opts: MultipartOptions
+  ): Promise<MultipartReadableStream | null>;
+
+  put(stream: ReadableStream<Uint8Array>): Promise<BlobId>;
+
+  delete(id: BlobId): Promise<void>;
+}
+
+function generateBlobId(): BlobId {
+  const idBuffer = Buffer.alloc(32 + 8);
+  crypto.randomFillSync(idBuffer, 0, 32);
+  idBuffer.writeBigUint64BE(process.hrtime.bigint(), 32);
+  return idBuffer.toString("hex");
+}
+
+export class MemoryBlobStore implements BlobStore {
+  readonly #blobs = new Map<string, Uint8Array>();
+
+  get(
+    id: BlobId,
+    range?: InclusiveRange
+  ): Promise<ReadableStream<Uint8Array> | null>;
+  get(
+    id: BlobId,
+    ranges: InclusiveRange[],
+    opts: MultipartOptions
+  ): Promise<MultipartReadableStream | null>;
+  async get(
+    id: BlobId,
+    ranges?: InclusiveRange | InclusiveRange[],
+    opts?: MultipartOptions
+  ): Promise<ReadableStream<Uint8Array> | MultipartReadableStream | null> {
+    const blob = this.#blobs.get(id);
+    if (blob === undefined) return null;
+    if (Array.isArray(ranges)) {
+      assert(opts !== undefined);
+      return createMultipartArrayReadableStream(blob, ranges, opts);
+    } else {
+      return createArrayReadableStream(blob, ranges);
+    }
+  }
+
+  async put(stream: ReadableStream<Uint8Array>): Promise<BlobId> {
+    const id = generateBlobId();
+    const buffer = await arrayBuffer(stream);
+    const blob = new Uint8Array(buffer);
+
+    // Store blob, making sure we're storing a new key
+    assert(!this.#blobs.has(id));
+    this.#blobs.set(id, blob);
+
+    return id;
+  }
+
+  async delete(id: BlobId): Promise<void> {
+    this.#blobs.delete(id);
+  }
+}
+
+export class FileBlobStore implements BlobStore {
+  readonly #root: string;
+
+  constructor(root: string) {
+    this.#root = path.resolve(root);
+  }
+
+  #idFilePath(id: BlobId) {
+    const filePath = path.join(this.#root, id);
+    return filePath.startsWith(this.#root) ? filePath : null;
+  }
+
+  get(
+    id: BlobId,
+    range?: InclusiveRange
+  ): Promise<ReadableStream<Uint8Array> | null>;
+  get(
+    id: BlobId,
+    ranges: InclusiveRange[],
+    opts: MultipartOptions
+  ): Promise<MultipartReadableStream | null>;
+  async get(
+    id: BlobId,
+    ranges?: InclusiveRange | InclusiveRange[],
+    opts?: MultipartOptions
+  ): Promise<ReadableStream<Uint8Array> | MultipartReadableStream | null> {
```

Comment:
I'm probably misunderstanding these types, but should this be `Promise<ReadableStream<Uint8Array> | null> | MultipartReadableStream>`?